src/v/cluster/archival_metadata_stm.cc

```diff
@@ -851,6 +852,42 @@ archival_metadata_stm::get_segments_to_cleanup() const {
     using lw_segment_meta = cloud_storage::partition_manifest::lw_segment_meta;
     std::vector<lw_segment_meta> backlog = _manifest->lw_replaced_segments();
 
+    // Make sure that 'replaced' list doesn't have any references to active
+    // segments. This is a protection from the data loss.
```

Comment:
Comment could be a bit more explicit: "This should not happen, but protects us from data loss in cases where bugs elsewhere"